File: impl/go/derivation.go
// Package imgdrv is a content-addressed IR for reproducible build
// descriptions.
//
// Describe a build, compute its store paths, serialize it, and get bytes
// identical to what Nix emits.
//
// This is the Go implementation, and its job in this project is to be the
// FALSIFICATION TEST for the theory in docs/theory.md: Go has no sum types, no
// higher-kinded types and minimal generics, so if the first-order signature
// needs more than finite products, Go is where it breaks. README.md records
// what it actually cost, which is the experimental result.
package imgdrv

import "sort"

// The named string types are not decoration. A store path, a hex digest and an
// output name are all strings, and confusing them is the class of bug this
// project has already paid for once: a derivation's own path and the hash by
// which it is known as an input are both 64-character strings, and swapping
// them yields a plausible wrong answer rather than an error.
//
// Go's defined types catch that at compile time, exactly as Rust's newtypes do
// and unlike Python's erased NewType. This is the one row of the typing table
// where Go is as strong as Rust.

// StorePath is an absolute path in the store, e.g. /nix/store/<32 chars>-hello.
type StorePath string

// Sha256Hex is a sha256 digest as 64 lowercase hex characters.
type Sha256Hex string

// OutputName is the name of one output of a derivation: out, dev, lib, ...
type OutputName string

// Output is one output of a derivation.
//
// HashAlgo and Hash are empty for an ordinary derivation. When they are set the
// derivation is FIXED-OUTPUT: it declares its result in advance, so its
// identity comes from the declared hash rather than from how it is built.
//
// HashAlgo is a string and not the HashAlgo enum-alike in edsl.go, on purpose:
// parsing has to be TOTAL over whatever real Nix wrote, so the wire type stays
// permissive and strictness lives on the construction side.
type Output struct {
	Name     OutputName
	Path     StorePath
	HashAlgo string
	Hash     string
}

// Fixed reports whether this output declares its content hash in advance.
func (o Output) Fixed() bool { return o.HashAlgo != "" }

// InputDrv is a dependency on specific outputs of another derivation.
//
// Depending on dev alone is a real and common case, so the set of needed output
// names belongs to the EDGE rather than to the target.
type InputDrv struct {
	Path    StorePath
	Outputs []OutputName
}

// EnvEntry is one environment variable.
//
// Go has no tuple type, so what is (string, string) in Python and Rust needs a
// named struct here. One line of cost, recorded because the point of this
// implementation is to count them.
type EnvEntry struct {
	Key   string
	Value string
}

// Derivation is a build description: the seven fields of the Derive(...) form.
//
// Field order is the serialization order and is load-bearing. See
// docs/spec/canonical.md.
type Derivation struct {
	Outputs   []Output
	InputDrvs []InputDrv
	InputSrcs []StorePath
	System    string
	Builder   string
	Args      []string
	Env       []EnvEntry
}

// OutputNames returns the names of this derivation's own outputs.
//
// Used when masking: an env entry is blanked when its KEY is one of these,
// never when an output path merely appears inside some value.
func (d Derivation) OutputNames() map[OutputName]bool {
	names := make(map[OutputName]bool, len(d.Outputs))
	for _, o := range d.Outputs {
		names[o.Name] = true
	}
	return names
}

// FixedOutput returns the fixed output and true, if this is a fixed-output
// derivation.
//
// Go has no Option, so a two-case sum becomes a (value, bool) pair. That is the
// idiomatic encoding and it is fine here; it stops being fine when the Option
// has to survive being stored in a struct, which is what Build.Outputs runs
// into. See edsl.go.
func (d Derivation) FixedOutput() (Output, bool) {
	for _, o := range d.Outputs {
		if o.Fixed() {
			return o, true
		}
	}
	return Output{}, false
}

// Name returns the derivation name, from the name environment variable.
//
// Every derivation carries one, and it is what output store names are built
// from.
func (d Derivation) Name() string {
	for _, e := range d.Env {
		if e.Key == "name" {
			return e.Value
		}
	}
	return ""
}

// Equal reports whether two derivations are structurally equal.
//
// Go has no derivable structural equality: a struct containing a slice is not
// comparable with ==, so this is hand-written for every type that has one.
// reflect.DeepEqual would do it in one line, at the cost of dropping to
// runtime-typed comparison, which is exactly what this implementation is
// supposed to avoid. Four Equal methods is the price; it is recorded in
// README.md rather than hidden.
func (d Derivation) Equal(other Derivation) bool {
	if d.System != other.System || d.Builder != other.Builder {
		return false
	}
	if !equalSlice(d.Args, other.Args) || !equalSlice(d.InputSrcs, other.InputSrcs) {
		return false
	}
	if !equalSlice(d.Env, other.Env) || !equalSlice(d.Outputs, other.Outputs) {
		return false
	}
	if len(d.InputDrvs) != len(other.InputDrvs) {
		return false
	}
	for i := range d.InputDrvs {
		if !d.InputDrvs[i].Equal(other.InputDrvs[i]) {
			return false
		}
	}
	return true
}

// Equal reports whether two edges are structurally equal.
func (i InputDrv) Equal(other InputDrv) bool {
	return i.Path == other.Path && equalSlice(i.Outputs, other.Outputs)
}

// equalSlice compares two slices of comparable elements.
//
// This is the one place generics earn their keep here, and they are the weak
// kind: a type parameter constrained to comparable, with an identical body for
// every element type. Exactly the case "write code, not types" allows.
func equalSlice[T comparable](a, b []T) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// sortedCopy returns a sorted copy of a slice of ordered values.
func sortedCopy[T ~string](in []T) []T {
	out := append([]T(nil), in...)
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

File: impl/rust/src/derivation.rs
//! The derivation types.
//!
//! These are the carrier of the whole crate: everything else parses into them,
//! serializes out of them, or hashes them.
//!
//! The newtypes are not decoration, and here they are load-bearing in a way
//! Python's `NewType` cannot be. A store path, a hex digest and an output name
//! are all strings at runtime, and confusing them is exactly the class of bug
//! this project has already paid for once: a derivation's own path and the hash
//! by which it is known as an input are both 64-character strings, and swapping
//! them yields a plausible wrong answer rather than an error. In Python that
//! confusion is caught only by `mypy`; here it does not typecheck at all, which
//! is one row of the typing table in `README.md`.
//!
//! Note what is deliberately NOT strongly typed: `Output::hash_algo` is a
//! `String`, not the [`HashAlgo`](crate::HashAlgo) enum. The wire type has to
//! round-trip whatever real Nix wrote, so parsing must be TOTAL; strictness
//! belongs on the construction side, in [`crate::edsl`].

use std::collections::BTreeSet;
use std::fmt;

/// A macro for the string newtypes, so the boilerplate is written once.
macro_rules! string_newtype {
    ($(#[$meta:meta])* $name:ident) => {
        $(#[$meta])*
        #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
        pub struct $name(String);

        impl $name {
            /// Wrap a string. Deliberately explicit: this is the one place the
            /// distinction between these types can be lost.
            pub fn new(s: impl Into<String>) -> Self {
                Self(s.into())
            }

            /// The underlying string.
            pub fn as_str(&self) -> &str {
                &self.0
            }

            /// Consume and return the underlying string.
            pub fn into_string(self) -> String {
                self.0
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(&self.0)
            }
        }

        impl From<&str> for $name {
            fn from(s: &str) -> Self {
                Self(s.to_owned())
            }
        }

        impl From<String> for $name {
            fn from(s: String) -> Self {
                Self(s)
            }
        }

        impl AsRef<str> for $name {
            fn as_ref(&self) -> &str {
                &self.0
            }
        }
    };
}

string_newtype! {
    /// An absolute path in the store, e.g. `/nix/store/<32 chars>-hello`.
    StorePath
}

string_newtype! {
    /// A sha256 digest as 64 lowercase hex characters.
    Sha256Hex
}

string_newtype! {
    /// The name of one output of a derivation: `out`, `dev`, `lib`, ...
    OutputName
}

/// One output of a derivation.
///
/// `hash_algo` and `hash` are empty for an ordinary derivation. When they are
/// set the derivation is FIXED-OUTPUT: it declares its result in advance, so
/// its identity comes from the declared hash rather than from how it is built.
/// That is the escape hatch that lets a build fetch from the network while
/// staying reproducible.
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct Output {
    /// The output's name: `out`, `dev`, `lib`, ...
    pub name: OutputName,
    /// The store path this output will occupy.
    pub path: StorePath,
    /// Empty, or the ingestion method and algorithm (`sha256`, `r:sha256`).
    /// A `String` and not an enum on purpose: parsing must be total.
    pub hash_algo: String,
    /// Empty, or the declared hash as lowercase hex.
    pub hash: String,
}

impl Output {
    /// Whether this output declares its content hash in advance.
    pub fn fixed(&self) -> bool {
        !self.hash_algo.is_empty()
    }
}

/// A dependency on specific outputs of another derivation.
///
/// Depending on `dev` alone is a real and common case, so the set of needed
/// output names belongs to the EDGE rather than to the target.
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct InputDrv {
    /// The `.drv` path of the derivation depended on.
    pub path: StorePath,
    /// The outputs of it that are actually needed.
    pub outputs: Vec<OutputName>,
}

/// A build description: the seven fields of the `Derive(...)` form.
///
/// Field order is the serialization order and is load-bearing. See
/// `docs/spec/canonical.md`.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct Derivation {
    /// This derivation's own outputs, sorted by name when canonical.
    pub outputs: Vec<Output>,
    /// Edges to other derivations, sorted by store path when canonical.
    pub input_drvs: Vec<InputDrv>,
    /// Store paths used directly as sources, sorted when canonical.
    pub input_srcs: Vec<StorePath>,
    /// The platform to build on, e.g. `x86_64-linux`.
    pub system: String,
    /// The program to run.
    pub builder: String,
    /// Arguments to the builder. Order is the MEANING and is never sorted.
    pub args: Vec<String>,
    /// Environment entries, sorted by key when canonical.
    pub env: Vec<(String, String)>,
}

impl Derivation {
    /// The names of this derivation's own outputs.
    ///
    /// Used when masking: an env entry is blanked when its KEY is one of these,
    /// never when an output path merely appears inside some value.
    pub fn output_names(&self) -> BTreeSet<&str> {
        self.outputs.iter().map(|o| o.name.as_str()).collect()
    }

    /// The fixed output, if this is a fixed-output derivation.
    pub fn fixed_output(&self) -> Option<&Output> {
        self.outputs.iter().find(|o| o.fixed())
    }

    /// The derivation name, from the `name` environment variable.
    ///
    /// Every derivation carries one, and it is what output store names are
    /// built from.
    pub fn name(&self) -> &str {
        self.env
            .iter()
            .find(|(k, _)| k == "name")
            .map(|(_, v)| v.as_str())
            .unwrap_or("")
    }
}
